lib/ide/file-manager/mock-server/src/main.rs

```diff
@@ -0,0 +1,228 @@
+                let reply_text   = serde_json::from_value::<String>(reply_json).unwrap();
+                let reply_msg    = Message::text(reply_text);
+                if self.socket.write_message(reply_msg).is_err() {
+                    return;
+                }
+            }
+            // ignore non-text messages.
+        }
+    }
+}
+
+
+
+// ==========
+// == Main ==
+// ==========
+
+fn main() {
+    let port    = utils::env::parse_var_or(PORT_VAR, DEFAULT_PORT);
+    let address = format!("127.0.0.1:{}", port);
+    let server  = TcpListener::bind(&address).unwrap();
+    println!("Listening on {}", address);
+    for stream in server.incoming() {
+        spawn(move || {
+            println!("Got a new connection");
+            let callback = |_req: &Request, response: Response| {
+                Ok(response)
+            };
+            let socket = accept_hdr(stream.unwrap(), callback).unwrap();
+            let mut server = Handler { socket };
+            server.run();
+        });
+    }
+}
+
+
+
+// =============
+// == IsCall  ==
+// =============
+
+/// All methods supported by this server should implement the trait.
+pub trait IsCall {
+    /// Type of value returned on a successful call.
+    type Result: Serialize;
+
+    /// Perform the "actual" work for this call.
+    fn realize(&self) -> CallResult<Self::Result>;
+}
+
+
+
+// ===========
+// == Call  ==
+// ===========
+
+/// Possible call requests. Members should wrap structs implementing `IsCall`.
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+#[serde(tag = "method", content = "params", rename_all = "camelCase")]
+enum Call {
+    CopyFile(CopyFile),
+    Exists(Exists),
+    Read(Read),
+    Write(Write),
+    // TODO [mwu]: add more methods, as needed
+}
+
+
+
+// ==============
+// == Methods  ==
+// ==============
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+struct CopyFile {from:PathBuf, to:PathBuf}
+impl IsCall for CopyFile {
+    type Result = ();
+    fn realize(&self) -> CallResult<()> {
+        match std::fs::copy(&self.from,&self.to) {
+            Ok(_)  => Ok(()),
+            Err(e) => Err(e.to_string())
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+struct Exists {path:PathBuf}
+impl IsCall for Exists {
+    type Result = bool;
+    fn realize(&self) -> CallResult<bool> {
+        Ok(self.path.exists())
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+struct Read {path:PathBuf}
+impl IsCall for Read {
+    type Result = String;
+    fn realize(&self) -> CallResult<String> {
+        std::fs::read_to_string(&self.path).map_err(|e| e.to_string())
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+struct Write {path:PathBuf, contents:String}
+impl IsCall for Write {
+    type Result = ();
+    fn realize(&self) -> CallResult<()> {
+        std::fs::write(&self.path,&self.contents).map_err(|e| e.to_string())
+    }
+}
```

Comment:
It should be done as macro. Together with generation of `enum Call`. It will be much shorter and will allow for fancy things, like automatic derivation of `From` impls.